src/logging.c

```diff
@@ -99,49 +148,96 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
+          "to control the stream and the buffering of log messages.\n");
       }
     } else {
       fprintf(
-        stderr, "Error getting env. variable "
+        stderr, "Error getting environment variable "
         "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Error getting environment variable RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s", ret_str);
+      return RCUTILS_RET_ERROR;
+    }
+
+    // Set the default output stream for all severities to stderr so that errors
+    // are propagated immediately.  The user can choose to set the output stream
+    // to stdout by setting the RCUTILS_CONSOLE_USE_STDOUT environment
+    // variable to 1.
+    enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
+      "use stdout");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_ZERO:
+        g_output_stream = stderr;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_output_stream = stdout;
+        break;
     }
 
-    // Check the environment variable for colorized output
-    const char * colorized_output;
-    ret_str = rcutils_get_env("RCUTILS_COLORIZED_OUTPUT", &colorized_output);
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_CONSOLE_BUFFERED_STREAM.  With an empty environment variable, use the
+    // default of the stream.  With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
+      "buffered");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+        break;
+      case RCUTILS_GET_ENV_ZERO:
+        if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", error_string);
+          return RCUTILS_RET_ERROR;
+        }
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", error_string);
+          return RCUTILS_RET_ERROR;
+        }
```

Comment:
How about de-duplicating these two blocks and choosing the argument to `setvbuf` based on `ZERO` / `ONE`?

Reply:
The issue is that we also have to be careful about "default", since the default is different based on whether we are using `stdout` or `stderr`.  The code would have to go something like:

```
switch (retval) {
    case ERROR:
        return;
    case ONE:
        arg = _IONBF;
    case ZERO:
        arg = _IOLBF;
}

if (retval != DEFAULT) {
    setvbuf(g_output_stream, NULL, arg, 0);
    // error checking
}
```

I don't really think that is better, but if you think we should change to that, I can.

Reply:
How about without any switch? E.g.:

```
if (ERRO == retval) {
  return;
}
if (ZERO == retval || ONE == retval) {
  auto arg = ONE == retval ? _IONBF : _IOLBF;
  setvbuf(g_output_stream, NULL, arg, 0);

Reply:
This is done now.